poppy/wfe.py

```diff
@@ -299,3 +300,343 @@ def get_opd(self, wave, units='meters'):
+        a[int(npix/2)+1:npix, 0] = sign*a[1:int(npix/2), 0][::-1]
+        a[int(npix/2)+1:npix, int(npix/2)+1:npix] = sign*np.rot90(a[1:int(npix/2), 1:int(npix/2)], 2)
+        a[int(npix/2)+1:npix, 1:int(npix/2)] = sign*np.rot90(a[1:int(npix/2), int(npix/2)+1:npix], 2)
+        
+        # remove any overall phase resulting from the zero-frequency component
+        a[0, 0] = 0.0
+        
+        return a
+    
+    def rand_turbulent(self, npix):
+        """ Returns a complex-valued random number array of shape (npix, npix)
+        with the symmetry required for a turbulent phase screen.
+        
+        Parameters
+        -----------------
+        npix : int
+            Number of pixels.
+        
+        References
+        -------------------
+        Eq. (63) in J. A. Fleck Jr, J. R. Morris, and M. D. Feit,
+        Appl. Phys. 10, 129 (1976).
+        """
+        
+        # create real-valued random numbers with required symmetry
+        a = self.rand_symmetrized(npix, 1)
+        b = self.rand_symmetrized(npix, -1)
+        
+        # create complex-valued random number with required variance
+        c = (a + 1j*b)/np.sqrt(2.0)
+        
+        return c
+    
+    def power_spectrum(self, wave, L0=None, l0=None, kind='Kolmogorov'):
+        """ Returns the spatial power spectrum.
+        
+        Parameters
+        -----------------
+        wave : wavefront object
+            Wavefront to calculate the power spectrum for.
+        
+        L0 : float
+            The outer scale of the turbulent eddies (m).
+        
+        l0 : float
+            The inner scale of the turbulent eddies (m).
+        
+        kind : string
+            The type of the power spectrum, must be one of 'Kolmogorov',
+            'Tatarski', 'van Karman', 'Hill'.
+        
+        References
+        -------------------
+        G. Gbur, J. Opt. Soc. Am. A 31, 2038 (2014).
+        
+        R. Frehlich, Appl. Opt. 39, 393 (2000).
+        """
+        
+        if not any(kind==item for item in ['Kolmogorov', 'Tatarski', 'van Karman', 'Hill']):
+            raise AttributeError('Kind of power spectrum not correctly defined.')
+        
+        Cn2 = self.get_Cn2(wave.wavelength)
+        coordinates = wave.coordinates()
+        npix = coordinates[0].shape[0]
+        pixelscale = wave.pixelscale.to(u.m/u.pixel)
+        
+        q = np.fft.fftfreq(npix, d=pixelscale.to(u.m/u.pixel))*2.0*np.pi
+        qx, qy = np.meshgrid(q, q)
+        
+        q2 = (qx**2 + qy**2)/u.pixel**2
+        if kind=='van Karman':
+            if L0 is not None:
+                q2 += 1.0/L0.to(u.m)**2
+            else:
+                raise AttributeError('If van Karman kind of turbulent phase \
+                                     screen is chosen, the outer scale L_0 \
+                                     must be provided.')
+        q2[0, 0] = np.inf # this is to avoid a possible error message in the next line
```

Comment:
I don't follow this part; why is it helpful to set `q2[0,0]` to infinity? What error message does it prevent?

Reply:
q2[0, 0] is zero (except for van Karman). The next line tries to divide by q2, so there would be a "division by zero" error. The value for q2[0, 0] is irrelevant really, because the random array at this position also has been set to 0.